src/libNode/Node.cpp

```diff
@@ -2408,6 +2427,199 @@ bool Node::ProcessDoRejoin(const bytes& message, unsigned int offset,
+
+  return true;
+}
+
+bool Node::ProcessNewShardGuardNetworkInfo(const bytes& message,
+                                           unsigned int offset,
+                                           const Peer& from) {
+  LOG_MARKER();
+
+  if (!GUARD_MODE) {
+    LOG_GENERAL(
+        WARNING,
+        "Not in guard mode. Unable to update shard guard network info.");
+    return false;
+  }
+
+  uint64_t dsEpochNumber;
+  Peer shardGuardNewNetworkInfo;
+  uint64_t timestamp;
+  PubKey shardGuardPubkey;
+
+  if (!Messenger::GetNodeNewShardGuardNetworkInfo(
+          message, offset, dsEpochNumber, shardGuardNewNetworkInfo, timestamp,
+          shardGuardPubkey)) {
+    LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum,
+              "Messenger::GetNodeNewShardGuardNetworkInfo failed.");
+    return false;
+  }
+
+  if (from.GetIpAddress() != shardGuardNewNetworkInfo.GetIpAddress()) {
+    LOG_CHECK_FAIL("IP Address",
+                   shardGuardNewNetworkInfo.GetPrintableIPAddress(),
+                   from.GetPrintableIPAddress());
+    return false;
+  }
+
+  if (m_mediator.m_selfKey.second == shardGuardPubkey) {
+    LOG_GENERAL(INFO,
+                "[update shard guard] Node to be updated is current node. No "
+                "update needed.");
+    return false;
+  }
+
+  uint64_t currentDSEpochNumber =
+      m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1;
+
+  if (dsEpochNumber != currentDSEpochNumber) {
+    LOG_GENERAL(
+        WARNING,
+        "Update of shard guard network info failure  - dsepoch in message: "
+            << dsEpochNumber
+            << " does not match current dsepoch: " << currentDSEpochNumber);
+    return false;
+  }
+
+  if (!Guard::GetInstance().IsNodeInShardGuardList(shardGuardPubkey)) {
+    LOG_GENERAL(WARNING,
+                "Update of shard guard requested by non-shard guard node, So "
+                "ignore the request");
+    return false;
+  }
+  // I am lookup node
+  if (LOOKUP_NODE_MODE) {
+    m_mediator.m_ds->UpdateShardNodeNetworkInfo(shardGuardNewNetworkInfo,
+                                                shardGuardPubkey);
+    if (!BlockStorage::GetBlockStorage().PutShardStructure(
+            m_mediator.m_ds->m_shards, 0)) {
+      LOG_GENERAL(WARNING, "BlockStorage::PutShardStructure failed");
+    }
+  }
+  // I am sharded node and requestor is also from my shard
+  else if (m_mediator.m_ds->m_mode == DirectoryService::IDLE) {
+    if (!IsShardNode(shardGuardPubkey)) {
+      LOG_GENERAL(WARNING, "PubKey of sender "
+                               << from
+                               << " does not match any of my shard members");
+      return false;
+    }
+    // update requestor's ( ShardGuard ) new IP
+    lock_guard<mutex> g(m_mutexShardMember);
+
+    unsigned int indexOfShardNode;
+    for (indexOfShardNode = 0;
+         indexOfShardNode < Guard::GetInstance().GetNumOfShardGuard();
```

Comment:
Since `GetNumOfShardGuard()` calls a non-inline function that must also acquire a mutex, it might be more efficient to put `GetNumOfShardGuard()` result into a const value instead.

Reply:
Will be removing the usage of this now.